File: 404-Sum-of-Left-Leaves.cpp
//Problem: Given a binary tree
//Return: Find the sum of all left leaves

#include <iostream>

 // Definition for a binary tree node.
 struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(0), right(0) {}
 };
 
class Solution {
public:
    int sumOfLeftLeaves(TreeNode* root) {
        if(!root)
        {
            return 0;
        }

        int total = 0;

        if((*root).left)
        {
            if((*this).children((*root).left))
            {
                total += (*this).sumOfLeftLeaves((*root).left);
            }
            else
            {
                total += (*((*root).left)).val;
            }
        }

        if((*root).right)
        {
            total += (*this).sumOfLeftLeaves((*root).right);
        }
    }

    bool children(TreeNode* node)
    {
        return(bool((*node).left) or bool((*node).right));
    }

};



int main()
{
    std::cout << "Test";
    return 0;
}

File: 404-Sum-of-Left-Leaves.py
# Problem: Given a binary tree
# Return: Find the sum of all left leaves


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def sumOfLeftLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return(0)
        return(self.DSF(root))
        
    def DSF(self,node):
        total = 0
        left = node.left
        right = node.right
        
        if left:
            if self.check_kids(left):
                total += self.DSF(left)
            else:
                total += left.val
                
        if right:
            total += self.DSF(right)
        
        return(total)
        
    def check_kids(self,node):
        return(node.left or node.right)